Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+
+private:
+  Index nNodes=0;
+  Index iNode=0;
+  std::vector<Data> data;
+
+
+  /*=== Getters ===*/
+public:
+  Data* GetData() {return data.data();}
+  const Data* GetData() const {return data.data();}
+
+  /*=== Outputting ===*/
+protected:
+  void toPythonString_fromto(std::ostream& output, Index from, Index to) const {
+    output  << "[";
+    for(Index i=from; i<to; ){
+      output << data[i];
+      if(++i<to) output << ", ";
+    }
+    output << "]";
+  }
+
+public:
+  NdFlattener(void) {}
+
+protected:
+  void allocate(){
+    data.reserve(nNodes);
+  }
+
+  void reset_iNode(){
+    iNode = 0;
+  }
+
+  /*=== Construct from 'recursive function' ===*/
+protected:
+  template<typename f_type>
+  void count_f(f_type f){
+    nNodes += f.first;
+  }
+
+  template<typename f_type>
+  void set_f(f_type f, bool refresh){
+    Index nChild = f.first;
+    for(Index iChild=0; iChild<nChild; iChild++){
+      data[iNode] = f.second(iChild);
+      iNode++;
+    }
+  }
+  
+  /*=== Construct with Allgatherv ===*/
+protected:
+  template<typename MPI_Environment_type>
+  void count_g(MPI_Environment_type const& mpi_env,
+         Index** Nodes_all,
+         CurrentLayer const* local_version,
+         int const* displs, int const* ones)
+  { 
+    assert( Nodes_all[0]==nullptr);
+    Nodes_all[0] = new Index[mpi_env.size];
+    nNodes = 0;
+    // gather numbers of nodes in the current layer from all processes
+    mpi_env.MPI_Allgatherv( &(local_version->nNodes), 1, mpi_env.mpi_index, Nodes_all[0], ones, displs, mpi_env.mpi_index, mpi_env.comm );
+    for(int r=0; r<mpi_env.size; r++){
+      nNodes += Nodes_all[0][r];
+    }
+  }
+
+  template<typename MPI_Environment_type>
+  void set_g(MPI_Environment_type const& mpi_env,
+         Index** Nodes_all,
+         CurrentLayer const* local_version )
+  { 
+
+
+    int* Nodes_all_0_as_int = new int[mpi_env.size];
+    int* Nodes_all_0_cumulated = new int[mpi_env.size+1];
+    Nodes_all_0_cumulated[0] = 0;
+    for(int r=0; r<mpi_env.size; r++){
+      Nodes_all_0_cumulated[r+1] = Nodes_all_0_cumulated[r] + Nodes_all[0][r];
+      Nodes_all_0_as_int[r] = Nodes_all[0][r];
+    }
+
+    mpi_env.MPI_Allgatherv( local_version->data.data(), Nodes_all[0][mpi_env.rank], mpi_env.mpi_data, data.data(), Nodes_all_0_as_int, Nodes_all_0_cumulated, mpi_env.mpi_data, mpi_env.comm );
+    delete[] Nodes_all_0_as_int;
+    delete[] Nodes_all_0_cumulated;
+  }
+
+
+  /*== Access to data ==*/
+  // Calling the following functions is a bit strange, because if you have only one level,
+  // you do not really benefit from NdFlattener's functionality.
+public:
+  Index size() const { // should not be called by recursion, is incorrect in lower layers!
+    return nNodes;
+  }
+  /*! \brief Data look-up, non-const version.
+   */
+  Data& operator[](Index i0) {
+    return data[i0];
+  }
+  /*! \brief Data look-up, const version.
+   */
+  const Data& operator[](Index i0) const {
+    return data[0];
```

Comment:
looks suspicious

Reply:
Good spot! There were no tests for `NdFlattener<1>`, let alone for `const NdFlattener`. So I extended the test set a bit.